src/Dscp/DB/SQLite/Queries.hs

```diff
@@ -140,6 +149,134 @@ getStudentTransactions student = do
         where      student_addr = ?
     |]
 
+getProvenStudentTransactionsSince :: DBM m => Id Student -> UTCTime -> m [(MerkleProof PrivateTx, [PrivateTx])]
+getProvenStudentTransactionsSince studentId sinceTime = do
+    transaction $ do
+        -- Contains `(tx, idx, blockId)` map.
+        txsBlockList <- getTxsBlockMap
+
+        -- Bake `blockId -> [(tx, idx)]` map.
+        let txsBlockMap = groupToAssocWith (_tibBlockId, _tibTx) txsBlockList
+
+        forM txsBlockMap $ \(blockId, transactions) -> do
+            blockData <- getBlockData blockId
+
+            let indices = _twiTxIdx <$> transactions & Set.fromList
+            let bodies  = _twiTx    <$> transactions
+            let tree    = _bdTree blockData
+            let pruned  = mkMerkleProof tree indices
+
+            return (pruned, bodies)
+
+  where
+    groupToAssocWith :: Eq k => Ord k => (a -> k, a -> v) -> [a] -> [(k, [v])]
+    groupToAssocWith (key, value)
+        -- The `groupBy` should return `[NonEmpty a]`.
+        = map      (\every@ (selected : _) -> (key selected, map value every))
+        . groupBy  ((==) `on` key)
+        . sortWith  key
```

Comment:
You map values three times using `key` -- with `sortWith`, `groupBy` and `map` itself. Isn't it faster to first convert to a `(k,v)` list and then transform into the `(k,[v])`?